File: KoshiSolvers/SolverEulerMethod.cs
﻿using System.Collections.Generic;
namespace KoshiSolvers
{
    public class EulerSolver: Solver
    {
        // Constructors
        public EulerSolver(string _Name, BehaviorOfSolver _Behaviour) : base(_Name ,_Behaviour) { }
        
        // Methods
        public override List<Point> Solve(InitialValueProblem Task)
        {
            double FunctionValue, StepSize = Task.H;
            int i = 1;
            List<Point> Solution = new List<Point>();

            Solution.Add(new Point(Task.T0, Task.Y0));
            FunctionValue = Task.CountFunctionValue(Solution[0].X, Solution[0].Y);

            while (Solution[i - 1].X < Task.T)
            {
                Solution.Add(new Point(Solution[i - 1].X + StepSize, Solution[i - 1].Y + StepSize * FunctionValue));
                FunctionValue = Task.CountFunctionValue(Solution[i].X, Solution[i].Y);

                // случаи непопадания на границу отрезка
                if ((Solution[i].X + StepSize) > Task.T)
                {
                    switch (Behavior)
                    {
                        case BehaviorOfSolver.FinishAtTheLeftBorder:

                            StepSize = Task.T - (StepSize * i + Solution[0].X);

                            Solution.Add(new Point(Solution[i].X + StepSize, Solution[i].Y + StepSize * FunctionValue));
                            break;

                        case BehaviorOfSolver.FinishAfterLeftBorder:
                            Solution.Add(new Point(Solution[i].X + StepSize, Solution[i].Y + StepSize * FunctionValue));
                            break;

                        case BehaviorOfSolver.FinishBeforeLeftBorder:
                            break;
                    }
                    break;
                }
                i++;
            }
            return Solution;
        }
    }    
}

File: C_Plus_Plus_Projects--2-4.1/SolverEulerMethod.cpp
#include "SolverEulerMethod.h"
#include <cmath>
#include <iostream>

void SolverEulerMethod::SolverKoshiTask(const TaskKoshi &Task)
{
	double FunctionValue;
	double StepSize = Task.Geth();
	unsigned int i = 1;

	Point.X[0] = Task.Gett0();
	Point.Y[0] = Task.Gety0();
	FunctionValue = Task.CountFunctionValue(Point.X[0],Point.Y[0]);

	while (Point.X[i] <= Task.GetT())
	{
		Point.X[i] = Point.X[i - 1] + StepSize;
		Point.Y[i] = Point.Y[i - 1] + StepSize * FunctionValue;
		FunctionValue = Task.CountFunctionValue(Point.X[i], Point.Y[i]);
		
		// случаи непопадания на границу отрезка
		if ((Point.X[i] + StepSize) > Task.GetT())
		{
			switch (Behavior) 
			{
			case BehaviorOfSolver::NoneBehavior:
				break;

			case BehaviorOfSolver::FinishAtTheLeftBorder: 

				StepSize = Task.GetT() - (StepSize * i + Point.X[0]);

				Point.X[i + 1] = Point.X[i] + StepSize;
				Point.Y[i + 1] = Point.Y[i] + StepSize * FunctionValue;
				break;
			
			case BehaviorOfSolver::FinishAfterLeftBorder: 

				Point.X[i + 1] = Point.X[i] + StepSize;
				Point.Y[i + 1] = Point.Y[i] + StepSize * FunctionValue;
				break;

			case BehaviorOfSolver::FinishBeforeLeftBorder:
				break;
			}
			break;
		}
		i++;
	}
}

